torch/_subclasses/functional_tensor.py

```diff
@@ -0,0 +1,259 @@
+from torch.utils._python_dispatch import return_and_correct_aliasing, TorchDispatchMode
+
+not_implemented_log = torch._logging.getArtifactLogger(__name__, "not_implemented")
+
+
+class _ToFunctionalTensor(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        return FunctionalTensor(x)
+
+    @staticmethod
+    def backward(ctx, grad_outputs):
+        raise RuntimeError(
+            "Attempted to backprop from a functional wrapper to its original tensor. This is unsupported behavior."
+        )
+
+
+class FunctionalTensor(torch.Tensor):
+    """
+    Functional tensors represent tensors that will remove mutations
+    from a program. If you perform a mutable operation on a functional tensor,
+    it will re-dispatch to the functional variant of that operation.
+
+    Historically, functionalization is implemented in C++ in the dispatcher.
+    This class is a lightweight python shim around the C++ functionalization logic.
+
+    FunctionalTensor is required to be used with a corresponding FunctionalTensorMode active.
+    It doesn't bother defining a __torch_dispatch__, because it relies
+    on using the mode for dispatch (which can properly handle factory functions).
+    """
+
+    elem: torch.Tensor
+    # Indicates to our torch_dispatch dispatching infra that
+    # this is an "infra" mode with lower dispatching precedence.
+    _mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
+
+    def __new__(cls, elem):
+        assert torch._is_functional_tensor(elem)
+        out = torch.Tensor._make_wrapper_subclass(
+            # TODO: right now, _make_wrapper_subclass's dynamic shape interaction is not great.
+            # Calling the overload that has kwargs causes us to go down the first overload path,
+            # which will **always** specialize sizes.
+            # We should probably eventually fix this so that the first overload can just handle dynamic shapes.
+            cls,
+            elem.shape,  # sizes
+            elem.stride(),  # strides
+            elem.storage_offset(),  # storage_offset
+            None,  # memory_format
+            elem.dtype,  # dtype
+            elem.layout,  # layout
+            elem.device,  # device
+            False,  # pin_memory
+            elem.requires_grad,  # requires_grad
+            "sizes",  # dispatch_sizes_strides_policy
+        )
+        out.elem = elem
+        return out
+
+    # Need to disable default torch_function. Why?
+    # Default torch_function will always wrap outputs into a subclass if they aren't already a subclass.
+    # We actually.. don't want to do this sometimes, see Note [FunctionalTensorMode inputs are sometimes plain tensors]
+    __torch_function__ = torch._C._disabled_torch_function_impl
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        unrecognized_types = [
+            t
+            for t in types
+            if not issubclass(t, torch._subclasses.FakeTensor)
```

Comment:
When do you want to allow subclasses of FakeTensor?

Reply:
Yeah you're right, this was probably too preemptive. Will remove.